### ISSUE TITLE
XPath functions for image handling

### ISSUE THREAD

**Issue by pr-apes** [speedata/xts#10]
@pgundlach,

it think some XPath functions are required for image handling (some are PDF specific).

* `sd:number-of-pages()`.
* `sd:aspectratio()`.
* `sd:imagewidth()`.
* `sd:imagewidth()`.

One comment about page number in layout functions, please allow variables directly (not requiring extra `concat()`).

I mean, the same way, `sd:aspectratio($pdfdoc)` is valid, `sd:aspectratio($pdfdoc:$pg:artbox)` should be available.

Functions for page dimensions (such as `sd-pagewidth()`) are also handy to manipulate images (to set their final size on the output page).

Many thanks for your help.

**Comment by pgundlach** [comment]
Since `$foo:$bar` is invalid XPath, I could allow a sequence as the argument instead of a file name:

`sd:number-of-pages( ($filename,$page,$box) )`

But then again, this is not far from `sd:number-of-pages( concat($filename,':',$page,':',$box) )` but would lead to less confusion.

So I am undecided on this.

**Comment by pgundlach** [comment]
Good point. I was thinking about this, but still not fixed yet:

`sd:imagefunction(  "filename" , "optional unit of measure")`

where "filename" can either be a string such as `ocean.pdf`  or a sequence `( "filename", "page number", "box" )`  , for example `("ocean.pdf", 2, "mediabox")`. Optional arguments can be left out by using an empty sequence: `("ocean.pdf", (), "mediabox")` which selects the default value.

**Comment by pgundlach** [comment]
OK, next release will have the image functions like this:

```
sd:imagefunction(<filename>[, <page-digit>], [<box>] [,<measure-unit>])
```

`sd:aspect-ratio()` will not accept a unit, but otherwise will be the same.

Thank you as always for the valuable input!

**Comment by pgundlach** [comment]
Implemented with d34ad433ccb99a24ec5174f7c7ecb8bb77a3aebd
